### PR TITLE
Added currency rates backups functionality

### PR DESCRIPTION
**Pull request by chyzas**
Backups are stored in Elasticsearch documents.

### CODE REVIEW

File: Command/UpdateEsCommand.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        try {
+            $data = $this->getContainer()->get('ongr_currency_exchange.currency_rates_service');
+            $data->reloadRates();
+            $output->writeln(sprintf('<info>Currency rates updated</info>'));
+        } catch (\Exception $e) {
+            $output->writeln(sprintf('<error>Error ocurred during update, check the logs!</error>'));
```
Comment (mindaugasbarysas) [inline Command/UpdateEsCommand.php]: What kind of exceptions could have occurred? I'm not sure if catching all of them is acceptable, especially if you don't really handle them afterwards. You also seem to make an assumption that whatever threw an exception has it logged somewhere, which might not be the case. So my suggestion would be to either rethrow that exception after writing that output line, or discard of the try catch block altogether.

File: Currency/Drivers/OpenExchangeRatesDriver.php [hunk 2]
```diff
@@ -25,17 +25,31 @@ class OpenExchangeRatesDriver implements CurrencyDriverInterface
     private $appId;
 
     /**
+     * @return string
```
Comment (mindaugasbarysas) [inline Currency/Drivers/OpenExchangeRatesDriver.php]: Methods should follow after properties.

File: Resources/config/services.yml [hunk 3]
```diff
@@ -22,5 +22,17 @@ services:
...
+            - @ongr_currency_exchange.currency_rates_service
+            - %ongr_currency_exchange.default_currency%
+
+   #Currency exchange drivers
+    #European central bank driver
+    ongr_currency_exchange.ecb_driver:
+        class: %ongr_currency_exchange.ecb_driver.class%
+    #Open exhange rates driver
+    ongr_currency_exchange.open_exchange_driver:
+        class: %ongr_currency_exchange.open_exchange_driver.class%
```
Comment (mindaugasbarysas) [inline Resources/config/services.yml]: Missing newline at the end of file.

File: Service/CurrencyRatesService.php [hunk 4]
```diff
@@ -80,48 +79,118 @@ public function getRates()
...
+        $search = $repository->createSearch();
+        $sort = new Sort('created_at', Sort::ORDER_DESC);
+        $search->addSort($sort);
+        $query = new MatchAllQuery();
+        $search->addQuery($query);
+        $search->setSize(1);
+        try {
+            $results = $repository->execute($search, Repository::RESULTS_ARRAY);
+        } catch (Missing404Exception $e) {
+            $this->logger && $this->logger->notice('Failed to execute query. Please check ES configuration');
```
Comment (mindaugasbarysas) [inline Service/CurrencyRatesService.php]: :+1: this is good.

File: Tests/Unit/DependencyInjection/ONGRCurrencyExchangeExtensionTest.php [hunk 5]
```diff
@@ -31,40 +31,17 @@ public function getTestLoadCurrencyData()
         $container1->setDefinition('ongr_currency_exchange.twig.price_extension', new Definition());
         $container1->setDefinition('stash.memcache', new Definition());
         $container1->setDefinition('my_service', new Definition());
```
Comment (mindaugasbarysas) [inline Tests/Unit/DependencyInjection/ONGRCurrencyExchangeExtensionTest.php]: Why container1? are there more? In any case, I think that naming scheme {something}{1..n} should be avoided.

File: DependencyInjection/ONGRCurrencyExchangeExtension.php [hunk 6]
```diff
@@ -31,7 +32,57 @@ public function load(array $configs, ContainerBuilder $container)
...
+                [
+                    new Reference($driver),
+                    new Reference(sprintf('es.manager.%s', $config['es_manager'])),
+                    new Reference($config['cache']),
+                ]
+            );
+            $def->addMethodCall('setLogger', [new Reference('logger')]);
+            $def->addTag('monolog.logger', ['channel' => 'ongr_currency']);
+            $container->setDefinition('ongr_currency_exchange.currency_rates_service', $def);
+        } else {
```
Comment (martiis) [inline DependencyInjection/ONGRCurrencyExchangeExtension.php]: else case is not needed here

File: Service/CurrencyRatesService.php [hunk 7]
```diff
@@ -80,48 +79,120 @@ public function getRates()
...
-            $this->reloadRates();
+        $this->rates = $this->getRatesFromBackup();
+        if (isset($this->rates)) {
+            return $this->rates;
+        }
+
+        $this->rates = $this->reloadRates();
+        if (isset($this->rates)) {
+            return $this->rates;
         } else {
```
Comment (martiis) [inline Service/CurrencyRatesService.php]: Else case is not needed here

File: Service/CurrencyRatesService.php [hunk 8]
```diff
@@ -80,48 +79,120 @@ public function getRates()
...
+
+        if (count($results)) {
+            foreach ($results[0]['rates'] as $data) {
+                $rates[$data['name']] = $data['value'];
+            }
+            $this->logger && $this->logger->notice('Rates returned from ES. Cache updated.');
+            $this->updateRatesCache($rates);
+
+            return $rates;
+        } else {
```
Comment (martiis) [inline Service/CurrencyRatesService.php]: Else case is not needed here also

**Pull request merged**